### PR TITLE
ghidra: improve instruction string and bytes feature extraction

### PR DESCRIPTION
**Pull request by mike-hunhoff**
### Checklist

- [X] No CHANGELOG update needed

- [X] No new tests needed

- [X] No documentation update needed

### CODE REVIEW

File: capa/features/extractors/ghidra/insn.py [hunk 1]
```diff
@@ -245,24 +236,10 @@ def extract_insn_string_features(fh: FunctionHandle, bb: BBHandle, ih: InsnHandl
...
-            ref = insn.getAddress(i)
-        # strings are also referenced dynamically via pointers & arrays, so we need to deref them
-        if insn.getOperandType(i) == dyn_addr:
-            ref = insn.getAddress(i)
-            dat = getDataAt(ref)  # type: ignore [name-defined] # noqa: F821
-            if dat and dat.isPointer():
-                ref = dat.getValue()
-
-    if ref != insn.getAddress():
-        ghidra_dat = getDataAt(ref)  # type: ignore [name-defined] # noqa: F821
```
Comment (mike-hunhoff) [inline capa/features/extractors/ghidra/insn.py:263]: `ref` may be `None` which causes a `RuntimeError` exception.

**Review by mr-tz — APPROVED** [review]
two questions, but overall looks good

File: capa/features/extractors/ghidra/helpers.py [hunk 2]
```diff
@@ -275,3 +275,31 @@ def dereference_ptr(insn: ghidra.program.database.code.InstructionDB):
...
+                ptr_value = data.getValue()
+
+                if ptr_value is None:
+                    break
+
+                to_addr = ptr_value
+            else:
+                break
+
+        yield to_addr
```
[the remark this answers is not in the document]
Comment (mike-hunhoff) [reply capa/features/extractors/ghidra/helpers.py:305]: is the yield only per reference intended here?

Yes - the intention is to follow nested points until data (as defined by Ghidra) is reached or `max_depth` is reached resulting in a single yield per memory reference from an instruction.

where we expect multiple references from an instruction?

Yes -  an instruction may have multiple references from it. Here we restrict our search to memory references that reference data (as defined by Ghidra).

**Comment by williballenthin** [comment]
i love the algorithms and methods that are being created here for capa + ghidra! i hope they serve as good documentation for the future

File: capa/features/extractors/ghidra/helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (williballenthin) [reply capa/features/extractors/ghidra/helpers.py:305]: perhaps symbolic analysis might resolve a set of candidate references for an indirect read/write like `mov [eax], 1`?

also perhaps ghidra supports an architecture with a complex memory to memory copy? (just guessing).

**Review by mr-tz — APPROVED** [review]
👍

**Comment by mike-hunhoff** [comment]
@colton-gabertan when you have a moment can you review and see if I missed anything obvious with these changes?

File: capa/features/extractors/ghidra/insn.py (hunk 1 above)
Comment (colton-gabertan) [reply capa/features/extractors/ghidra/insn.py:268]: Thank you for the catch and refactoring. I've updated it for the latest changes and tested it. Looks good to me.

**Review by colton-gabertan — APPROVED** [review]

**Pull request merged**